akka-docs-dev/rst/scala/stream-customize.rst

```diff
@@ -149,14 +149,175 @@ Using DetachedStage
+
+In our case we use the :class:`ReadPreferred` read condition which has the exact semantics which we need to implement
+our preferring merge – it pulls elements from the preferred input port if there are any available, otherwise reverting
+to pulling from the secondary inputs. The context object passed into the state function allows us to interact with the
+connected streams, for example by emitting an ``element``, which was just pulled from the given ``input``, or signalling
+completion or errors to the merges downstream stage.
+
+The state function must always return the next behaviour to be used when an element should be pulled from its upstreams,
+we use the special :class:`SameState` object which signals :class:`FlexiMerge` that no state transition is needed.
+
+Implementing Zip-like merges
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+More complex fan-in junctions may require not only multiple States but also sharing state between those states.
+As :class:`MergeLogic` is allowed to be stateful, it can be easily used to hold the state of the merge junction.
+
+We now implement the equivalent of the built-in ``Zip`` junction by using the property that a the MergeLogic can be stateful
+and that each read is followed by a state transition (much like in Akka FSM or ``Actor#become``).
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#fleximerge-zip-states
+
+The above style of implementing complex flexi merges is useful when we need fine grained control over consuming from certain
+input ports. Sometimes however it is simpler to strictly consume all of a given set of inputs. In the ``Zip`` rewrite below
+we use the :class:`ReadAll` read condition, which behaves slightly differently than the other read conditions, as the element
+it is emitting is of the type :class:`ReadAllInputs` instead of directly handing over the pulled elements:
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#fleximerge-zip-readall
+
+Thanks to being handed a :class:`ReadAllInputs` instance instead of the elements directly it is possible to pick elements
+in a type-safe way based on their input port.
+
+Connecting your custom junction is as simple as creating an instance and connecting Sources and Sinks to its ports
+(notice that the merged output port is named ``out``):
+
+.. includecode:: code/docs/stream/FlexiDocSpec.scala#fleximerge-zip-connecting
+
+.. _flexi-merge-completion-handling-scala:
+
+Completion handling
+^^^^^^^^^^^^^^^^^^^
+Completion handling in :class:`FlexiMerge` is defined by an :class:`CompletionHandling` object which can react on
+completion and error signals from its upstream input ports. The default strategy is to remain running while at-least-one
+upstream input port is still running (i.e. has not signalled completion or error).
```

Comment:
well, if you expressed that you only want to read from port X, and port X happens to be closed then it does not remain running?